iavl_nodedb.go

```diff
@@ -73,68 +107,416 @@ func (ndb *nodeDB) SaveNode(node *IAVLNode) {
 		cmn.PanicSanity("Shouldn't be calling save on an already persisted node.")
 	}
 
-	// Save node bytes to db
+	// Don't overwrite nodes.
+	// This is here because inner nodes are overwritten otherwise, losing
+	// version information, due to the version not affecting the hash.
+	// Adding the version to the hash breaks a lot of things, so this
+	// seems like the best solution for now.
+	if ndb.Has(node.hash) {
+		return
+	}
+
+	// Save node bytes to db.
 	buf := new(bytes.Buffer)
 	if _, err := node.writeBytes(buf); err != nil {
 		cmn.PanicCrisis(err)
 	}
-	ndb.batch.Set(node.hash, buf.Bytes())
+	ndb.batch.Set(ndb.nodeKey(node.hash), buf.Bytes())
+
 	node.persisted = true
 	ndb.cacheNode(node)
+}
+
+// Has checks if a hash exists in the database.
+func (ndb *nodeDB) Has(hash []byte) bool {
+	key := ndb.nodeKey(hash)
+
+	if ldb, ok := ndb.db.(*dbm.GoLevelDB); ok {
+		exists, err := ldb.DB().Has(key, nil)
+		if err != nil {
+			cmn.PanicSanity("Got error from leveldb: " + err.Error())
+		}
+		return exists
+	}
+	return len(ndb.db.Get(key)) != 0
+}
 
-	// Re-creating the orphan,
-	// Do not garbage collect.
-	delete(ndb.orphans, string(node.hash))
-	delete(ndb.orphansPrev, string(node.hash))
+// SaveBranch saves the given node and all of its descendants. For each node
+// about to be saved, the supplied callback is called and the returned node is
+// is saved. You may pass nil as the callback as a pass-through.
+//
+// Note that this function clears leftNode/rigthNode recursively and calls
+// hashWithCount on the given node.
+func (ndb *nodeDB) SaveBranch(node *IAVLNode, cb func(*IAVLNode)) {
+	if node.hash == nil {
```

Comment:
This is my issue with version.

version is set in https://github.com/tendermint/iavl/blob/feature/historical-queries/iavl_orphaning_tree.go#L75-L80 that is only in the callback

the first time a node is saved, we calculate the hashes recursively here. (with version=0)